test/common/network/udp_listener_impl_test.cc

```diff
@@ -172,33 +189,72 @@ TEST_P(UdpListenerImplTest, LargeDatagramRecvmmsg) {
   EXPECT_EQ(2, listener_->packetsDropped());
 }
 
-// Test a large datagram that gets dropped using recvmsg.
-TEST_P(UdpListenerImplTest, LargeDatagramRecvmsg) {
+TEST_P(UdpListenerImplTest, LimitNumberOfReadsPerLoop) {
   setup();
+  const uint64_t num_packets_per_read =
+      Api::OsSysCallsSingleton::get().supportsMmsg() ? NUM_DATAGRAMS_PER_MMSG_RECEIVE : 1u;
+
+  size_t num_packets_expected_per_loop{32u};
+  // These packets should be read in more than 3 loops.
+  const std::string payload1(10, 'a');
+  for (uint64_t i = 0; i < 2 * num_packets_expected_per_loop; ++i) {
+    client_.write(payload1, *send_to_addr_);
+  }
+  const std::string last_piece("bbb");
+  client_.write(last_piece, *send_to_addr_);
 
-  ON_CALL(override_syscall_, supportsMmsg()).WillByDefault(Return(false));
-
-  // This will get dropped.
-  const std::string first(4096, 'a');
-  client_.write(first, *send_to_addr_);
-  const std::string second("second");
-  client_.write(second, *send_to_addr_);
-  // This will get dropped.
-  const std::string third(4096, 'b');
-  client_.write(third, *send_to_addr_);
+  EXPECT_CALL(listener_callbacks_, onReadReady()).Times(testing::AtLeast(3u));
+  EXPECT_CALL(listener_callbacks_, numPacketsExpectedPerEventLoop())
+      .WillRepeatedly(Return(num_packets_expected_per_loop));
+  EXPECT_CALL(listener_callbacks_, onData(_))
+      .WillRepeatedly(Invoke([&](const UdpRecvData& data) -> void {
+        validateRecvCallbackParams(data, num_packets_per_read);
+        if (last_piece == data.buffer_->toString()) {
+          dispatcher_->exit();
+        }
+      }));
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
 
+  num_packets_received_by_listener_ = 0u;
+  num_packets_expected_per_loop = 0u;
+  std::string payload2(10, 'c');
+  // This packet should be read.
+  client_.write(payload2, *send_to_addr_);
   EXPECT_CALL(listener_callbacks_, onReadReady());
-  EXPECT_CALL(listener_callbacks_, onDatagramsDropped(_)).Times(AtLeast(1));
+  EXPECT_CALL(listener_callbacks_, numPacketsExpectedPerEventLoop())
+      .WillRepeatedly(Return(num_packets_expected_per_loop));
   EXPECT_CALL(listener_callbacks_, onData(_)).WillOnce(Invoke([&](const UdpRecvData& data) -> void {
-    validateRecvCallbackParams(
-        data, Api::OsSysCallsSingleton::get().supportsMmsg() ? NUM_DATAGRAMS_PER_MMSG_RECEIVE : 1u);
-    EXPECT_EQ(data.buffer_->toString(), second);
-
+    validateRecvCallbackParams(data, num_packets_per_read);
+    EXPECT_EQ(payload2, data.buffer_->toString());
     dispatcher_->exit();
   }));
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
 
+  if (!recvbuf_large_enough_) {
+    // If SO_RCVBUF failed to enlarge receive buffer to 4MB, the rest of test will likely to fail
```

Comment:
CI is failing on format check:

fix_spelling_pedantic...
./test/common/network/udp_listener_impl_test.cc:234:    // If SO_RCVBUF failed to enlarge receive buffer to 4MB, the rest of test will likely to fail
RCVBUF:

Reply:
done